python/ray/_private/accelerators/tpu.py

```diff
@@ -247,3 +191,181 @@ def set_current_process_visible_accelerator_ids(
+
+        if tpu_accelerator_type is not None:
+            ray_accelerator_type = tpu_accelerator_type_to_ray_accelerator_type(
+                tpu_accelerator_type
+            )
+            if ray_accelerator_type is None:
+                logger.info(
+                    "While trying to autodetect a TPU type, "
+                    f"received malformed accelerator_type: {tpu_accelerator_type}"
+                )
+
+        if ray_accelerator_type is None:
+            logging.info("Failed to auto-detect TPU type.")
+
+        return ray_accelerator_type
+
+    def postprocess_resources(resources: Mapping[str, float]):
+        """Apply TPU pod-specific postprocessing to input resources.
+
+        This will populate the TPU version and additional resources
+        used for TPU pod execution.
+
+        When running workloads on a TPU pod, we need a way to run
+        the same binary on every worker in the TPU pod.
+
+        See https://jax.readthedocs.io/en/latest/multi_process.html
+        for more information.
+
+        To do this in ray, we take advantage of custom resources. We
+        mark worker 0 of the TPU pod as a "coordinator" that identifies
+        the other workers in the TPU pod. We therefore need:
+        - worker 0 to be targetable.
+        - all workers in the TPU pod to have a unique identifier consistent
+        within a TPU pod.
+
+        So assuming we want to run the following workload:
+
+        @ray.remote
+        def my_jax_fn():
+            import jax
+            return jax.device_count()
+
+        We could broadcast this on a TPU pod (e.g. a v4-16) as follows:
+
+        @ray.remote(resources={"tpu-v4-16"})
+        def run_jax_fn(executable):
+            # Note this will execute on worker 0
+            tpu_name = ray.util.accelerators.tpu.get_tpu_pod_name()
+            num_workers = ray.util.accelerators.tpu.get_tpu_num_workers()
+            tpu_executable = executable.options(resources={"TPU": 4, tpu_name: 1})
+            return [tpu_executable.remote() for _ in range(num_workers)]
+
+        Returns:
+            An optional list of strings representing:
+            - the TPU accelerator version, e.g. "TPU-V2", "TPU-V3",
+              "TPU-V4" if applicable, else None.
+            - the TPU pod name
+            - if worker 0, the TPU topology.
+        """
+        tpu_id = TPUAcceleratorManager.get_tpu_id()
+        worker_id = TPUAcceleratorManager.get_tpu_worker_id()
+        tpu_accelerator_type = TPUAcceleratorManager.get_tpu_accelerator_type()
+
+        if tpu_id and worker_id is not None and tpu_accelerator_type:
+            pod_resource_name = f"TPU-{tpu_accelerator_type}"
+            # Add the name of the TPU ID to the resource.
+            resources[tpu_id] = 1
+            # Only add in the TPU pod resource type to worker 0.
+            if worker_id == 0:
+                resources[pod_resource_name] = 1
```

Comment:
Who is going to use those resources, is it end-user or higher level libraries. I'm asking to see how hard it will be if we later migrate them to labels.

Reply:
Mostly by end users and probably a small handful of higher level libraries by Google. There will likely need to be a migration, but I expect it to be pretty contained